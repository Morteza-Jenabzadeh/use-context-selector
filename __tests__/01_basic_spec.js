import React, { useRef, useState, StrictMode } from 'react';

import {
  render, fireEvent, cleanup, screen,
} from '@testing-library/react';

import {
  createContext,
  useContextSelector,
} from '../src/index';

describe('basic spec', () => {
  afterEach(cleanup);

  it('counter', () => {
    const initialState = {
      count1: 0,
      count2: 0,
    };
    const context = createContext(null);
    const Counter1 = () => {
      const count1 = useContextSelector(context, (v) => v[0].count1);
      const setState = useContextSelector(context, (v) => v[1]);
      const increment = () => setState((s) => ({
        ...s,
        count1: s.count1 + 1,
      }));
      const renderCount = useRef(0);
      renderCount.current += 1;
      return (
        <div>
          <span>{count1}</span>
          <button type="button" onClick={increment}>+1</button>
          <span>{renderCount.current}</span>
        </div>
      );
    };
    const Counter2 = () => {
      const count2 = useContextSelector(context, (v) => v[0].count2);
      const renderCount = useRef(0);
      renderCount.current += 1;
      return (
        <div>
          <span>{count2}</span>
          <button type="button">+1</button>
          <span data-testid="counter2">{renderCount.current}</span>
        </div>
      );
    };
    const StateProvider = ({ children }) => {
      const [state, setState] = useState(initialState);
      return (
        <context.Provider value={[state, setState]}>
          {children}
        </context.Provider>
      );
    };
    const App = () => (
      <StrictMode>
        <StateProvider>
          <Counter1 />
          <Counter2 />
        </StateProvider>
      </StrictMode>
    );
    const { getAllByText, container } = render(<App />);
    expect(container).toMatchSnapshot();
    fireEvent.click(getAllByText('+1')[0]);
    expect(screen.getByTestId('counter2').textContent).toEqual('1');
    expect(container).toMatchSnapshot();
  });
});
