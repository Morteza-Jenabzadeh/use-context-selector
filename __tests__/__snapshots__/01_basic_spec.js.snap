// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic spec counter 1`] = `
<div>
  <div>
    <span>
      0
    </span>
    <button
      type="button"
    >
      +1
    </button>
    <span>
      1
    </span>
  </div>
  <div>
    <span>
      0
    </span>
    <button
      type="button"
    >
      +1
    </button>
    <span
      data-testid="counter2"
    >
      1
    </span>
  </div>
</div>
`;

exports[`basic spec counter 2`] = `
<div>
  <div>
    <span>
      1
    </span>
    <button
      type="button"
    >
      +1
    </button>
    <span>
      3
    </span>
  </div>
  <div>
    <span>
      0
    </span>
    <button
      type="button"
    >
      +1
    </button>
    <span
      data-testid="counter2"
    >
      1
    </span>
  </div>
</div>
`;
